Mathlib/Topology/Algebra/ProperAction.lean

```diff
@@ -0,0 +1,308 @@
+/-
+Copyright (c) 2024 Anatole Dedeker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anatole Dedeker, Etienne Marion, Florestan Martin-Baillon, Vincent Guirardel
+-/
+import Mathlib.Topology.Algebra.Group.Basic
+import Mathlib.Topology.Algebra.MulAction
+import Mathlib.Topology.Maps.Proper.Basic
+import Mathlib.Topology.Sequences
+
+/-!
+# Proper group action
+
+In this file we define proper action of a group on a topological space, and we prove that in this
+case the quotient space is T2. We also give equivalent definitions of proper action using
+ultrafilters and show the transfer of proper action to a closed subgroup.
+We give sufficent conditions on the topological space such that the action is properly discontinuous
+(see `ProperlyDiscontinuousSMul`) if and only if it is continuous in
+the first variable (see `ContinuousConstSMul`) and proper in the sense defined here.
+
+## Main definitions
+
+* `ProperSMul` : a group `G` acts properly on a topological space `X`
+  if the map `(g, x) ↦ (g • x, x)` is proper, in the sense of `IsProperMap`.
+
+## Main statements
+
+* `t2Space_of_ProperSMul`: If a group `G` acts properly on a topological space `X`,
+  then the quotient space is Hausdorff (T2).
+* `t2Space_of_properSMul_of_t2Group`: If a T2 group acts properly on a topological space,
+  then this topological space is T2.
+* `ProperlyDiscontinuousSMul_iff_ProperSMul`: If a discrete group acts on a T2 space `X` such that
+  `X × X` is compactly generated, then the action is properly discontinuous if and only if it is
+  continuous in the second variable and proper. This in particular true if `X` is locally compact
+  or first-countable.
+
+## Implementation notes
+
+Concerning `ProperlyDiscontinuousSMul_iff_ProperSMul`, this result should be the only one needed
+to link properly discontinuous and proper actions, however the notion of compactly generated space
+is not yet in Mathlib so it is used as an intermediate result to prove
+`WeaklyLocallyCompactSpace.isProperMap_iff_isCompact_preimage` and
+`SequentialSpace.isProperMap_iff_isCompact_preimage`.
+
+## References
+
+* [N. Bourbaki, *General Topology*][bourbaki1966]
+
+## Tags
+
+Hausdorff, group action, proper action
+-/
+
+open Filter Topology Set Prod
+
+/-- Proper group action in the sense of Bourbaki:
+the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+@[mk_iff]
+class ProperVAdd (G X : Type*) [TopologicalSpace G] [TopologicalSpace X] [AddGroup G]
+    [AddAction G X] : Prop where
+  /-- Proper group action in the sense of Bourbaki:
+  the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+  isProperMap_vadd_pair' : IsProperMap (fun gx ↦ ⟨gx.1 +ᵥ gx.2, gx.2⟩ : G × X → X × X)
+
+/-- Proper group action in the sense of Bourbaki:
+the map `G × X → X × X` is a proper map (see `isProperMap`). -/
+@[to_additive existing, mk_iff]
```

Comment:
I conjecture that
```suggestion
@[to_additive existing (attr := mk_iff)]
```
will save you from the `attribute [to_additive existing] properSMul_iff` line (but you should check! I'm not sure)

Reply:
What if you remove the `@[mk_iff]` in front of `ProperVAdd`?

Reply:
It doesn't work either. This time I get the following:
```
application type mismatch
  @properSMul_iff G X inst✝¹ inst✝ inst✝³
argument has type
  AddGroup G
but function has type
  ∀ [inst : Group G] [inst_1 : MulAction G X], ProperSMul G X ↔ IsProperMap fun gx ↦ (gx.1 • gx.2, gx.2)
```
in the proofs where I use `properSMul_iff` and there is a `to_additive` attribute.

Reply:
Okay what about you also add `attribute [to_additive existing] properSMul_iff` after the definition of `ProperSMul` and before you use `properSMul_iff`?

Reply:
Ok it seems to work, so in the end I removed `@[mk_iff]` in front of `ProperVAdd` instead of removing `attribute [to_additive existing] properSMul_iff`.